### PR TITLE
Request buffering middleware

### LINKED ISSUE #2043: Buffering client requests.
### Do you want to request a *feature* or report a *bug*?
Feature

### What did you do?
We have a Traefik instance set up as ingress controller in our Kubernetes cluster. Before we used Nginx as a proxy for our services.
It was configured to buffer client' requests before sending them to upstream/destination.
Based on access metrics from Nginx and self-reported ones from microservices we set up alerts to trigger when response time passes the defined threshold. It was working fine.
However Traefik doesn't buffer clients' requests and pipes them straight to client.
In case of clients with slow user connection and e.g. file upload request between traefik and microservice/destination can take very long (tens of seconds).
It is generally fine, as we don't have any control over users connection. Problem is with metrics as we are no longer able to tell whether microservice behaves correctly based on request time.
Our Apdex alerts don't work. 

Another drawback of not buffering requests is using more resources on microservices behind Traefik which work in thread-per-request model.

### What did you expect to see?

We can base view on microservice healthiness on metrics we get from Traefik / self-reported by microservice.

### What did you see instead?

Metrics are highly affected by clients connection speed.

### Proposed solutions
* allow requests buffering (AFAIK it is supported by https://github.com/vulcand/oxy)
* provide more information about request processing time in logs. (e.g. last byte sent to destination to last byte received from destination)

### PR DESCRIPTION
**Pull request by harnash**
### What does this PR do?

This approach allows better control over requests processing incoming into Traefik. It allows to specify buffer limits and maximum request size that will be processed by Traefik. I've decided to use global settings for all requests (which seems enough for us) but it may more convenient to specify those settings per frontend.

For more background about this change please see the linked issue and mentioned PR.

### Motivation

Fixes #2043 

### More

- [x] Added/updated tests
- [x] Added/updated documentation
- [x] Docker Provider updated
- [x] Rancher Provider updated
- [x] Marathon Provider updated
- [x] k8s Provider updated
- [x] Mesos Provider updated
- [x] ECS Provider updated
- [x] Consul Catalog Provider updated
- [x] KV Provider updated

### Additional Notes

This is related to PR https://github.com/containous/traefik/pull/2209

~This code will not test properly since Traefik uses forked vulcand/oxy dependency which does not have `buffer` package included (merge with upstream needed).~

### CODE REVIEW

**Comment by harnash** [comment]
@emilevauge I was thinking about something similar. Since it is still blocked I will try to move those settings into backend config.

**Comment by nmengin** [comment]
ping @harnash 

Did you try to move move the settings into backend config? Do you need some help?

**Comment by harnash** [comment]
@nmengin Hey. Sadly no. I had my hands full but I'll try to look at this this weekend

**Comment by harnash** [comment]
@nmengin I've just pushed f120a47 which moves configuration to backend config. I might deploy this on our dev cluster soon to test it but I'm not sure I can since to build it I need vulcand/oxy/buffering package which is missing in the containous' fork. Any advice how to deal with this?

**Comment by Juliens** [comment]
As we need to have the last vulcand/oxy, I think we must wait for https://github.com/containous/traefik/issues/2390

**Comment by Juliens** [comment]
The new oxy version is merged on master, can you rebase ?
Thx

**Comment by harnash** [comment]
It is rebased but tests will still be failing since fork is missing `oxy.buffering` package.

**Comment by ldez** [comment]
@harnash pb fixed

**Comment by harnash** [comment]
@nmengin I'll look into that

**Comment by ldez** [comment]
@harnash any news?

**Comment by harnash** [comment]
@ldez sorry I hadn't time to look at this yet but I'll schedule some time tomorrow.

**Comment by harnash** [comment]
I'm still working on it. It seems that I need to update Providers to accommodate new settings. It will take a bit of more work than expected.

**Comment by harnash** [comment]
@ldez I still need to update some providers - I'll ping you when it is ready for review

**Comment by ldez** [comment]
@harnash you must rebase to have the well fromed Mesos provider

**Comment by ldez** [comment]
@harnash if you have some problem with the vendoring, do a `dep ensure -v`.
https://github.com/containous/traefik/blob/master/CONTRIBUTING.md#setting-up-dependency-management

**Comment by ldez** [comment]
@harnash I will your vendoring.

**Comment by ldez** [comment]
@harnash due to some flaky tests, could you rebase again thanks 😃

**Comment by harnash** [comment]
@ldez I think I'm done with this PR unless more rebases are coming ;-)

File: provider/kubernetes/kubernetes.go [hunk 1]
```diff
@@ -295,6 +295,42 @@ func (p *Provider) loadIngresses(k8sClient Client) (*types.Configuration, error)
 					}
 				}
 
+				if service.Annotations[label.TraefikBackendBufferingEnabled] == "true" {
```
Comment (ldez) [inline provider/kubernetes/kubernetes.go]: could you create a dedicate method.

```go
func getBuffering(service *v1.Service) *types.Buffering {
 # ...
}
```

File: types/types.go [hunk 2]
```diff
@@ -50,6 +51,16 @@ type CircuitBreaker struct {
 	Expression string `json:"expression,omitempty"`
 }
 
+// Buffering holds request/response buffering configuration/
+type Buffering struct {
+	Enabled              bool   `json:"enabled"`
```
Comment (nmengin) [inline types/types.go]: Can you delete the field `enabled`? Usually, in Træfik, defining attribute of a structure allows activating the feature.

File: provider/docker/config_container.go [hunk 3]
```diff
@@ -214,6 +214,23 @@ func getHealthCheck(container dockerData) *types.HealthCheck {
 	}
 }
 
+func getBuffering(container dockerData) *types.Buffering {
+	enabled := label.GetBoolValue(container.Labels, label.TraefikBackendBufferingEnabled, false)
```
Comment (ldez) [inline provider/docker/config_container.go]: Could you remove the `label.TraefikBackendBufferingEnabled` and instead use `label.HasPrefix(...)`

Same for all providers.

File: templates/consul_catalog.tmpl [hunk 4]
```diff
@@ -34,6 +34,27 @@
     interval = "{{ $healthCheck.Interval }}"
   {{end}}
 
+  {{ $buffering := getBuffering $service.Attributes }}
+  {{if $buffering }}
+  [backends.backend-{{ $backendName }}.buffering]
+    enabled = {{ $buffering.Enabled }}
+    {{if $buffering.MaxRequestBodyBytes }}
```
Comment (nmengin) [inline templates/consul_catalog.tmpl]: You don't have to check if each attribute is present.
You can just check `buffering` in *all* the templates.

File: server/server.go [hunk 5]
```diff
@@ -1560,3 +1571,31 @@ func (s *Server) wrapHTTPHandlerWithAccessLog(handler http.Handler, frontendName
 	}
 	return handler
 }
+
+func (s *Server) buildBufferingMiddleware(handler http.Handler, config *types.Buffering) (http.Handler, error) {
+	var lb *buffer.Buffer
```
Comment (ldez) [inline server/server.go]: could you rewrite this method like that:
```go
func (s *Server) buildBufferingMiddleware(handler http.Handler, config *types.Buffering) (http.Handler, error) {
	log.Debugf("Setting up buffering: request limits: %d (mem), %d (max), response limits: %d (mem), %d (max) with retry: '%s'",
		config.MemRequestBodyBytes, config.MaxRequestBodyBytes, config.MemResponseBodyBytes,
		config.MaxResponseBodyBytes, config.RetryExpression)

	if len(config.RetryExpression) > 0 {
		return buffer.New(
			handler,
			buffer.MemRequestBodyBytes(config.MemRequestBodyBytes),
			buffer.MaxRequestBodyBytes(config.MaxRequestBodyBytes),
			buffer.MemResponseBodyBytes(config.MemResponseBodyBytes),
			buffer.MaxResponseBodyBytes(config.MaxResponseBodyBytes),
			buffer.Retry(config.RetryExpression),
		)
	}

	return buffer.New(
		handler,
		buffer.MemRequestBodyBytes(config.MemRequestBodyBytes),
		buffer.MaxRequestBodyBytes(config.MaxRequestBodyBytes),
		buffer.MemResponseBodyBytes(config.MemResponseBodyBytes),
		buffer.MaxResponseBodyBytes(config.MaxResponseBodyBytes),
	)
}
```

File: provider/kv/keynames.go [hunk 6]
```diff
@@ -15,6 +15,12 @@ const (
 	pathBackendServers                          = "/servers/"
 	pathBackendServerURL                        = "/url"
 	pathBackendServerWeight                     = "/weight"
+	pathBackendBufferingEnabled                 = "/buffering/enabled"
```
Comment (ldez) [inline provider/kv/keynames.go]: could you modify this section like that:
```go
const (
// ...
	pathBackendBuffering                        = "/buffering/"
	pathBackendBufferingMaxResponseBodyBytes    = pathBackendBuffering + "maxresponsebodybytes"
	pathBackendBufferingMemResponseBodyBytes    = pathBackendBuffering + "memresponsebodybytes"
	pathBackendBufferingMaxRequestBodyBytes     = pathBackendBuffering + "maxrequestbodybytes"
	pathBackendBufferingMemRequestBodyBytes     = pathBackendBuffering + "memrequestbodybytes"
	pathBackendBufferingRetryExpression         = pathBackendBuffering + "retryexpression"
// ...
)
```

File: provider/kubernetes/kubernetes_test.go [hunk 7]
```diff
@@ -510,6 +538,17 @@ func TestServiceAnnotations(t *testing.T) {
 					server("http://10.15.0.2:8080", weight(1))),
 				lbMethod("wrr"), lbSticky(),
 			),
+			backend("baz",
+				servers(
+					server("http://10.14.0.1:8080", weight(1)),
+					server("http://10.12.0.1:8080", weight(1))),
+				lbMethod("wrr"),
+				maxRequestBodyBytes(10485760),
```
Comment (ldez) [inline provider/kubernetes/kubernetes_test.go]:
```go
buffering(
	maxRequestBodyBytes(10485760),
	memRequestBodyBytes(2097152),
	maxResponseBodyBytes(10485760),
	memResponseBodyBytes(2097152),
	retrying("IsNetworkError() && Attempts() <= 2"),
),
```

File: provider/label/names.go [hunk 8]
```diff
@@ -80,6 +87,12 @@ const (
 	TraefikBackendLoadBalancerStickinessCookieName = Prefix + SuffixBackendLoadBalancerStickinessCookieName
 	TraefikBackendMaxConnAmount                    = Prefix + SuffixBackendMaxConnAmount
 	TraefikBackendMaxConnExtractorFunc             = Prefix + SuffixBackendMaxConnExtractorFunc
+	TraefikBackendBufferingEnabled                 = Prefix + SuffixBackendBufferingEnabled
```
Comment (ldez) [inline provider/label/names.go]:
```go
const (
// ...
    TraefikBackendBuffering                        = Prefix + SuffixBackendBuffering
// ...
)

**Comment by harnash** [comment]
I'll address those comments soon.

**Comment by ldez** [comment]
Due to some conflicts with other PRs and the fact your PR is old (we are sorry for the delay), we will fix your PR.

**Pull request merged**